pkg/backend/s3/s3.go

```diff
@@ -15,61 +15,57 @@
 package s3
 
 import (
-	"os"
+	"fmt"
 
 	"github.com/hashicorp/hcl/v2"
 	"github.com/hashicorp/hcl/v2/gohcl"
-	"github.com/pkg/errors"
-
 	"github.com/kinvolk/lokomotive/internal/template"
-	"github.com/kinvolk/lokomotive/pkg/backend"
 )
 
-type s3 struct {
+// Config represents the configuration of an S3 backend.
+type Config struct {
 	Bucket        string `hcl:"bucket"`
 	Key           string `hcl:"key"`
-	Region        string `hcl:"region,optional"`
+	Region        string `hcl:"region"`
 	AWSCredsPath  string `hcl:"aws_creds_path,optional"`
 	DynamoDBTable string `hcl:"dynamodb_table,optional"`
 }
 
-// init registers s3 as a backend.
-func init() {
-	backend.Register("s3", NewS3Backend())
-}
+// NewConfig creates a new Config and returns a pointer to it as well as any HCL diagnostics.
+func NewConfig(b *hcl.Body, ctx *hcl.EvalContext) (*Config, hcl.Diagnostics) {
+	diags := hcl.Diagnostics{}
+
+	c := &Config{}
 
-// LoadConfig loads the configuration for the s3 backend.
-func (s *s3) LoadConfig(configBody *hcl.Body, evalContext *hcl.EvalContext) hcl.Diagnostics {
-	if configBody == nil {
-		return hcl.Diagnostics{}
+	if b == nil {
+		return nil, diags
 	}
-	return gohcl.DecodeBody(*configBody, evalContext, s)
-}
 
-func NewS3Backend() *s3 {
-	return &s3{}
-}
+	if d := gohcl.DecodeBody(*b, ctx, c); len(d) != 0 {
+		diags = append(diags, d...)
+		return nil, diags
+	}
 
-// Render renders the Go template with s3 backend configuration.
-func (s *s3) Render() (string, error) {
-	return template.Render(backendConfigTmpl, s)
+	return c, diags
 }
 
-// Validate validates the s3 backend configuration.
-func (s *s3) Validate() error {
-	if s.Bucket == "" {
```

Comment:
In this case it seems that "validate" means "ensure a bucket and a key were provided". If that's all we do, why not enforce this via HCL?

As for HCL vs. other formats - that's a valid concern but a purely theoretical in this context. I'm fine with having a more complex validation but IMO we should only have it where it's necessary (e.g. when we need to check more than "was X specified?").

Reply:
> If that's all we do, why not enforce this via HCL?

Because IMO it doesn't enforce a thing, see #239 for more details.

>As for HCL vs. other formats - that's a valid concern but a purely theoretical in this context. I'm fine with having a more complex validation but IMO we should only have it where it's necessary (e.g. when we need to check more than "was X specified?").

This is important and not theoretical, as the test for required variables shouldn't require getting HCL parser, context etc etc. Just a struct and a validation function should be enough to test validation.

Reply:
It is theoretical because we don't support other config formats at the moment :shrug:

Anyway, I'll re-add the validation.

Reply:
Done.

Reply:
Thanks :)